src/derived/slot.rs

```diff
@@ -489,210 +468,131 @@ where
-                    self, other_id,
-                );
-                match self.register_with_in_progress_thread(db, runtime, other_id, waiting) {
-                    Ok(future) => {
-                        // Release our lock on `self.state`, so other thread can complete.
-                        std::mem::drop(state);
-
-                        let result = future.wait().unwrap_or_else(|| Cancelled::throw());
-                        return !result.cycle.is_empty() || result.value.changed_at > revision;
-                    }
-
-                    // Consider a cycle to have changed.
-                    Err(_) => return true,
+        // Do an initial probe with just the read-lock.
+        //
+        // If we find that a cache entry for the value is present
+        // but hasn't been verified in this revision, we'll have to
+        // do more.
+        loop {
+            match self.maybe_changed_since_probe(db, self.state.read(), runtime, revision_now) {
+                MaybeChangedSinceProbeState::Retry => continue,
+                MaybeChangedSinceProbeState::ChangedAt(changed_at) => return changed_at > revision,
+                MaybeChangedSinceProbeState::Stale(state) => {
+                    drop(state);
+                    return self.maybe_changed_since_upgrade(db, revision);
                 }
             }
+        }
+    }
 
-            QueryState::Memoized(memo) => memo,
-        };
-
-        if memo.revisions.verified_at == revision_now {
-            debug!(
-                "maybe_changed_since({:?}): {:?} since up-to-date memo that changed at {:?}",
-                self,
-                memo.revisions.changed_at > revision,
-                memo.revisions.changed_at,
-            );
-            return memo.revisions.changed_at > revision;
+    fn maybe_changed_since_probe<StateGuard>(
+        &self,
+        db: &<Q as QueryDb<'_>>::DynDb,
+        state: StateGuard,
+        runtime: &Runtime,
+        revision_now: Revision,
+    ) -> MaybeChangedSinceProbeState<StateGuard>
+    where
+        StateGuard: Deref<Target = QueryState<Q>>,
+    {
+        match self.probe(db, state, runtime, revision_now) {
+            ProbeState::Retry => MaybeChangedSinceProbeState::Retry,
+
+            ProbeState::Stale(state) => MaybeChangedSinceProbeState::Stale(state),
+
+            // If we know when value last changed, we can return right away.
+            // Note that we don't need the actual value to be available.
+            ProbeState::NoValue(_, changed_at)
+            | ProbeState::UpToDate(StampedValue {
+                value: _,
+                durability: _,
+                changed_at,
+            }) => MaybeChangedSinceProbeState::ChangedAt(changed_at),
+
+            // If we have nothing cached, then value may have changed.
+            ProbeState::NotComputed(_) => MaybeChangedSinceProbeState::ChangedAt(revision_now),
```

Comment:
Hm, this seems questionable to me. If we return `revision_now` here, than in `maybe_changed_since` we actually return `false` from `changed_at > revision` condition. That is, we treat the input as if it didn't change. 

It seems that this can actually lead to the bugs, when the query state flips from some state to non-computed, but triggering that should be tricky.

Reply:
Well, it's not really valid to call `maybe_change_since` with revision now. The meaning of "maybe changed since R" is "did this change in revision (R+1)..now", but it'd be good to document that. The reason for this is that we invoke `maybe_changed_since(Rv)` where `Rv` is the "revision in which the data was last verified". What we really want to know, then, is did something change *after* that revision (if it changed at the *start* of Rv, we've already taken that into account).

Reply:
Renamed to `maybe_changed_after` and documented this more clearly. Also added various debug assertions.